### DISCUSSION TITLE
Issues maintaining open stream

### DISCUSSION THREAD

**Discussion by Chew** [redouane59/twittered#362] [Q&A]
I have a stream in my bot, all seems to work well until it just... randomly stops inexplicably.

Here's some code:

```java
        TwitterClient twitterClient = new TwitterClient(TwitterCredentials.builder()
            .accessToken(PropertiesManager.getTwitterAccessToken())
            .accessTokenSecret(PropertiesManager.getTwitterAccessTokenSecret())
            .apiKey(PropertiesManager.getTwitterAPIKey())
            .apiSecretKey(PropertiesManager.getTwitterAPIKeySecret())
            .build());

        twitterClient.startFilteredStream(tweet -> ServerChannel.TWITTER.getAsChannel().sendMessage(
            "Check out the latest tweet from x - Follow us on there today!\nhttps://twitter.com/x/status/" + tweet.getId()
        ).queue());
```

However, after a while, I just get "Check out the latest tweet from x - Follow us on there today!
https://twitter.com/x/status/null" the key thing being null. Is there a way to make it reconnect to the stream? How did it disconnect? What can I do to fix it :(

**Comment by redouane59** [comment]
Hello @Chew , I'm sorry I'm currently on holiday. I'll try to check as soon as possible.

**Comment by redouane59** [comment]
Hey again @Chew , if it can help, here is how I use the stream function : 

```
  public void start() {

    try {
      System.out.println("start streaming... ");
      this.streamResponse = TWITTER_CLIENT.startFilteredStream(new IAPIEventListener() {
        @SneakyThrows
        @Override
        public void onStreamError(final int httpCode, final String error) {
          System.err.println("HttpCode : " + httpCode + " - error : " + error);
          boolean result = TWITTER_CLIENT.stopFilteredStream(streamResponse);
          System.err.println("stopping stream... " + result);
          Thread.sleep(1000L * 5 * 60);
          System.err.println("restarting after 5min sleep");
          start();
        }

        @Override
        public void onTweetStreamed(final Tweet tweet) {
          try {
            Optional<TweetContent> tweetContent = incomingTweetsManager.prepareAnswer(tweet);
            if (tweetContent.isPresent()) {
                // some code
            }
          } catch (Exception e) {
            System.err.println("error streaming tweet : " + e.getMessage());
            System.err.println(Arrays.toString(e.getStackTrace()));
          }
        }

        @Override
        public void onUnknownDataStreamed(final String json) {
          System.err.println("onUnknownDataStreamed " + json);
          boolean result = TWITTER_CLIENT.stopFilteredStream(streamResponse);
          System.err.println("stopping stream... " + result);
        }

        @Override
        public void onStreamEnded(final Exception e) {
          System.err.println("onStreamEnded " + e.getMessage() + " / " + e.getStackTrace()[0]);
          boolean result = TWITTER_CLIENT.stopFilteredStream(streamResponse);
          System.err.println("stopping stream... " + result);
        }
      });
    } catch (Exception e) {
      System.err.println("- error - " + e.getMessage());
      if (streamResponse != null) {
        boolean result = TWITTER_CLIENT.stopFilteredStream(streamResponse);
        System.err.println("catch stopping stream... " + result);
      }
    }
  }
```

**Reply by Chew** [reply]
Was finally able to give this a shot. Unfortunately, the "null" tweet still happens and it never restarts. I suppose I can just check for that instead.

**Comment by snicoll** [comment]
@redouane59 isn't the issue here that there isn't a feature to make sure the stream keeps running in case of network error? The `onStreamEnded` callback can happen for two reasons:

1. It ended up unexpectedly (network I/O error, etc)
2. The user called `stopFilteredStream` while incoming tweets were processed

What I am looking for is a way to handle 1. so that the stream is kept alive until I explicitly ask it to be closed.

**Reply by Greg-M73** [reply]
@snicoll 
Suggestion : you can use a boolean on your main process like a 'bClosing' that you set to true when you explicitly close the stream. If a network happen, test this boolean to decide whether you restart the stream or ignore the close callback ?

**Reply by snicoll** [reply]
Yup, that's an option. What I am wondering is whether implementing such a feature in this project rather than each of us having to deal with this would be an option.

**Reply by Greg-M73** [reply]
You're right :). It's been some time since my last pull request, but i think that it would be easy to do as described with the existing classes.. (sry, i have absolutely no time to do it myself).

**Comment by akuma0** [comment]
Hi guys !
Any update here ?

**Comment by snicoll** [comment]
@akuma0 please don't crosspost. You already asked this in an issue #439.
